pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,302 @@
+package virtdhcp
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"syscall"
+
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const (
+	DNSmasqExec = "/usr/sbin/dnsmasq"
+	DHCPPidFile = "/tmp/dhcp.pid"
+)
+
+var DHCPLeaseFile = "/var/run/kubevirt/dhcp.leases"
+var DHCPHostsFile = "/var/run/kubevirt/dhcp.hosts"
+
+type hostsDetails struct {
+	Mac   string
+	Lease int
+}
+
+type DNSmasqInstance struct {
+	hosts    map[string]hostsDetails
+	ipRange  [][]byte
+	confpath string
+	args     []string
+	cmd      *exec.Cmd
+}
+
+func NewDNSmasq() *DNSmasqInstance {
+	dnsMasq := new(DNSmasqInstance)
+	dnsMasq.hosts = make(map[string]hostsDetails)
+	dnsMasq.ipRange = make([][]byte, 2)
+	dnsMasq.args = []string{"-k", "-d", "--strict-order", "--bind-dynamic", "--no-resolv", "--dhcp-no-override", "--dhcp-authoritative",
+		"--except-interface=lo", "--dhcp-hostsfile=" + DHCPHostsFile, "--dhcp-leasefile=" + DHCPLeaseFile,
+		"--pid-file=" + DHCPPidFile}
+	OS = &OSHandler{}
+	return dnsMasq
+}
+
+// implement `Interface` in sort package.
+type sortByteArrays [][]byte
+
+func (b sortByteArrays) Len() int {
+	return len(b)
+}
+
+func (b sortByteArrays) Less(i, j int) bool {
+	// bytes package already implements Comparable for []byte.
+	switch bytes.Compare(b[i], b[j]) {
+	case -1:
+		return true
+	case 0, 1:
+		return false
+	default:
+		return false
+	}
+}
+
+func (b sortByteArrays) Swap(i, j int) {
+	b[j], b[i] = b[i], b[j]
+}
+
+type OSOps interface {
+	IsFileExist(path string) bool
+	getPidFromFile(file string) (int, error)
+	checkProcessExist(pid int) (bool, *os.Process)
+	killProcessIfExist(pid int) error
+	RecreateHostsFile() (*os.File, error)
+	executeCommand(string, []string) (*exec.Cmd, error)
+	readFromFile(file string) ([]byte, error)
+}
+
+type OSHandler struct{}
+
+var OS OSOps
+
+func (o *OSHandler) IsFileExist(path string) bool {
+	_, err := os.Stat(DHCPPidFile)
+	return !os.IsNotExist(err)
```

Comment:
This logic might produce some unexpected results. If an error is returned that does not match "NotExist", then it will appear like the file exists when it may or may not. 

maybe do something like this (pulled from pkg/ephemeral-disk-utils/utils.go)
```
        _, err := os.Stat(path)
        exists := false

        if err == nil {
                exists = true
        } else if os.IsNotExist(err) {
                err = nil
        }
        return exists, err
```

Reply:
Thanks